### ISSUE TITLE
How to generate mm39.address? How was mm10.genomeInfo$gapInfo made? How was the MASK (T/F) column created?

### ISSUE THREAD

**Issue by Pancreas-Pratik** [zwdzwd/sesameData#8]
Good morning @zwdzwd,

I hope all is well, and again, as usual ; ), thank you for all your help thus far! I have been working on and off trying to get the mm39 files made. I think I have reached a point where I may need your help.

------------------

**1. How to generate the `mm39.address`?
2. How to generate the `MASK (TRUE/FALSE)` column using the [MM285.mm39.manifest](http://zhouserver.research.chop.edu/InfiniumAnnotation/current/MM285/MM285.mm39.manifest.tsv.gz)?
3. How was the `mm10.genomeInfo$gapInfo` `GRanges` object made?**

-------------------

**Question 1 and 2** are sort-of together... For the mm39.address, I am puzzled on how to determine the MASK column. I tried to look for patterns in the [MM285.mm39.manifest](http://zhouserver.research.chop.edu/InfiniumAnnotation/current/MM285/MM285.mm39.manifest.tsv.gz), similarly to the + and - strand designation https://github.com/zwdzwd/sesame/issues/47#issuecomment-914672976, but was not successful.

Also an important part of question 1 is how are the control probes determined? I checked out the [MM285.mm39.manifest](http://zhouserver.research.chop.edu/InfiniumAnnotation/current/MM285/MM285.mm39.manifest.tsv.gz) compared to the `MM285.mm10.manifest`, and found that the `mm10` has 4216 control probes, and these were in a separate list (`MM285.address$controls`). Does this mean that `mm39` has none of these control probes? **Does it matter if `mm39` does or doesn't for the generation of beta values all the through pipeline to DMLs?**

-----------------------------

**For Question 3:** I was able to generate all the components of `mm39.genomeInfo` (ie. SeqInfo, cytoBand, gene2txn, txn2gene, txns) except I am not exactly sure how to generate the `gapInfo` `GRanges `object. I am trying to understand how the `mm10.genomeInfo$gapInfo` was made so that I could apply this same procedure to complete the `mm39.genomeInfo`.

These is what I think was done for the `mm10.genomeInfo$gapInfo`:

```
##gapInfo using cdsByOverlaps()
mm10.txdb <- GenomicFeatures::makeTxDbFromEnsembl(organism = "Mus musculus", release = 102)
GenomeInfoDb::seqlevelsStyle(mm10.txdb) = "UCSC"
cds <- cdsByOverlaps(x = mm10.txdb, ranges =  MM285.mm10.manifest, columns = c("TXNAME"))
cds_grl <- multisplit(cds, cds$TXNAME)
cds_combined <- unlist(range(cds_grl))
gapInfo.by.overlaps <- gaps(cds_combined)
length(gapInfo.by.overlaps)
[1] 10381

##gapInfo using cds()
cds <- cds(mm39.txdb, columns=c("TXNAME"))
cds_grl <- multisplit(cds, cds$TXNAME)
cds_combined <- unlist(range(cds_grl))
gapInfo <- gaps(cds_combined)
length(gapInfo)
[1] 21975
```

However, this does not match the `mm10.genomeInfo$gapInfo`. There are 574 ranges within the `mm10.genomeInfo$gapInfo`, and about roughly 20,000+ more in the the one I have generated using the above code.

Thank you very much in advance.

Sincerely,
Pratik

**Comment by Pancreas-Pratik** [comment]
Good afternoon @zwdzwd,

Here is the code I have so far:
```r
library(GenomicFeatures)
library(GenomicRanges)
library(GenomeInfoDb)
library(sesameData)
#adding %>% functionality without loading up tidyverse completely
`%>%` <- magrittr::`%>%`

########################################
#####Adding mm39 to MM285.probeInfo#####
########################################

##making MM285.probeInfo
curl::curl_download(url = 'http://zhouserver.research.chop.edu/InfiniumAnnotation/current/MM285/MM285.mm39.manifest.tsv.gz', destfile = '~/Downloads/MM285.mm39.manifest.tsv.gz')
system('gunzip -c ~/Downloads/MM285.mm39.manifest.tsv.gz > ~/Downloads/MM285.mm39.manifest.tsv')
MM285.mm39.manifest <- read.table(file = "~/Downloads/MM285.mm39.manifest.tsv", sep = "\t", header = TRUE)

##converting mapFlag to strand (16 for '-' and 0 for '+')
MM285.mm39.manifest$strand <- MM285.mm39.manifest$mapFlag
MM285.mm39.manifest$strand <- gsub(MM285.mm39.manifest$strand, pattern = "16",  replacement = "-")
MM285.mm39.manifest$strand <- gsub(MM285.mm39.manifest$strand, pattern = "0",  replacement = "+")
colnames(MM285.mm39.manifest)[1:3] <- c("seqnames", "start", "end")
chrOrder<-c(paste("chr",1:19,sep=""),"chrX","chrY","chrM")
MM285.mm39.manifest$seqnames <- factor(MM285.mm39.manifest$seqnames, levels = chrOrder)
rm(chrOrder)
MM285.mm39.manifest <- MM285.mm39.manifest[order(MM285.mm39.manifest$seqnames),]

rownames(MM285.mm39.manifest) <- MM285.mm39.manifest$probeID

mapped.probes.mm39 <- GenomicRanges::makeGRangesFromDataFrame(MM285.mm39.manifest, 
                                                               keep.extra.columns = FALSE, 
                                                               starts.in.df.are.0based = FALSE)
##Downloading MM285.probeInfo to add on mapped.probes.mm39
MM285.probeInfo <- sesameData::sesameDataGet('MM285.probeInfo')
mapped.probes.mm10 <- MM285.probeInfo$mapped.probes.mm10

MM285.probeInfo <- list(mapped.probes.mm10 = mapped.probes.mm10, mapped.probes.mm39 = mapped.probes.mm39)

#clearing some of the environment
rm(mapped.probes.mm10, mapped.probes.mm39, MM285.mm39.manifest)


#################################
#####Building genomeInfo.mm39####
#################################

#Downloading mm39 Ensembl Release 104
mm39.txdb <- GenomicFeatures::makeTxDbFromEnsembl(organism = "Mus musculus", release = 104)
mm39.txdb <- GenomeInfoDb::keepStandardChromosomes(mm39.txdb)
#changing chromosome names to UCSC format "1,2,3..." to "chr1, chr2, chr3..."
GenomeInfoDb::seqlevelsStyle(mm39.txdb) = "UCSC"


##making txns
txns <- GenomicFeatures::cds(x = mm39.txdb,  columns = c("CDSSTART","CDSEND", "TXNAME", "GENEID"))
txns = split(txns, unlist(txns$TXNAME))

##making txn2gene
txn2gene <- AnnotationDbi::select(mm39.txdb, keys = names(txns), columns = c("GENEID"), keytype = "TXNAME")
myList <- as.list(txn2gene$GENEID)
names(myList) <- txn2gene$TXNAME
txn2gene <- myList
rm(myList)

##making gene2txn
gene2txn <- data.frame(AnnotationDbi::select(mm39.txdb, keys = names(txns), columns = c("GENEID"), keytype = "TXNAME"))
library(org.Mm.eg.db)
ens2symb <- AnnotationDbi::select(x = org.Mm.eg.db, keys = gene2txn$GENEID, keytype="ENSEMBL", column=c("SYMBOL"))
ens2symb <- unique(ens2symb)
colnames(ens2symb)[1] <- 'GENEID'
gene2txn <- merge(ens2symb, gene2txn, by = 'GENEID', all = TRUE)
rm(ens2symb)
gene2txn$GENEID <- NULL
gene2txn <- dplyr::distinct(gene2txn)
table(duplicated(gene2txn))
gene2txn <- gene2txn %>% dplyr::group_by(SYMBOL) %>%
  dplyr::mutate(TXNAME = paste(TXNAME, collapse=",")) %>% dplyr::distinct()
gene2txn <- dplyr::distinct(gene2txn)
gene2txn <- na.omit(gene2txn)
table(duplicated(gene2txn))

gene2txn2 <- as.list(gene2txn$TXNAME)
names(gene2txn2) <- gene2txn$SYMBOL
gene2txn <- gene2txn2
rm(gene2txn2)
gene2txn <- lapply(gene2txn[], function(x) unlist(strsplit(x, split = ","))) 

#finishing txns
mcols(txns, level="within")[, "cdsStart"] <- mcols(txns, level="within")[, "CDSSTART"]
mcols(txns, level="within")[, "cdsEnd"] <- mcols(txns, level="within")[, "CDSEND"]
txns <- txns[, c("cdsStart", "cdsEnd")]

#making SeqInfo
seqInfo <- seqinfo(mm39.txdb)

##downloading/making cytoBand from UCSC
####visual check of CytoBand start and end matches SeqInfo
curl::curl_download(url = 'http://hgdownload.soe.ucsc.edu/goldenPath/mm39/database/cytoBandIdeo.txt.gz', destfile = '~/Downloads/cytoBandIdeo.txt.gz')
system('gunzip -c ~/Downloads/cytoBandIdeo.txt.gz > ~/Downloads/mm39_cytoBand.txt')
cytoBand <- read.table(file = '~/Downloads/mm39_cytoBand.txt', sep = "\t")
cytoBand[6] <- "bed"
##Downloading genomeInfo.mm10 to use cytoBand column names
genomeInfo.mm10 <- sesameData::sesameDataGet('genomeInfo.mm10')
##copying over cytoband columns names
colnames(cytoBand) <- names(genomeInfo.mm10$cytoBand)

##MISSING gapInfo

genomeInfo.mm39 <- list(seqInfo = seqInfo, 
                        #gapInfo =  gapInfo, 
                          cytoBand = cytoBand, 
                        gene2txn = gene2txn,
                        txn2gene = txn2gene, 
                        txns = txns)

#Cleanup
rm(seqInfo, 
   cytoBand, 
   #gapInfo, 
   gene2txn, 
   txn2gene,
   txns)
```

Really, any help would be appreciated as I am trying to move forward a project I am working on!

Looking forward to your response!

Sincerely,
Pratik

**Comment by zwdzwd** [comment]
technically the address file is most independent of genome assembly for mouse array. if I understand correctly the mm39.address the only difference are color and MASK column which might be (very slightly) adjusted based on mapping against mm39. All the rest should remain the same.

I was using the UCSC mm10 to determine gapInfo. I guess you could use something like

```r
library(rtracklayer)
sess = browserSession('UCSC')
genome(sess) = 'mm39'
cytoBand = getTable(ucscTableQuery(sess, table='cytoBand'))
seqnames = paste0('chr',c(1:19, 'X','Y','M'))

chromInfo = getChromInfoFromUCSC('mm39')
chromInfo = chromInfo[chromInfo$chrom %in% seqnames,]

seqlengths <- chromInfo$size[match(seqnames, chromInfo$chrom)]
seqInfo <- Seqinfo(seqnames, seqlengths)

gap = getTable(ucscTableQuery(sess, table='gap'))
gap = gap[gap$chrom %in% seqnames,]
gapInfo = sort(GRanges(as.vector(gap$chrom), IRanges(gap$chromStart+1, gap$chromEnd), seqinfo=seqinfo))
```

Hope this is useful. Sorry for the late response.

**Comment by Pancreas-Pratik** [comment]
Thank you so very much, sir. I understand you must be extremely busy. I have been really checking obsessively for your response since my project depends on this!

@zwdzwd, could you take the time to answer two other related questions, please?

1. I just want to double check I understand what you mean sir. You mean, I can essentially use the MASK and color column that is on the `mm10.address$ordering` data frame and use it with the mm39 probes?
2. How are the control probes determined in the MM285.address? (There is another data frame in the MM285.address$controls - I am thinking the mm39 does not have these probes as I do not see them since I don't see any with the 'ctl' prefix?) 

Really looking forward to your response!

Sincerely,
Pratik

**Comment by zwdzwd** [comment]
1. Yes. for the array data to be extracted all you need is the probe address on the chip and annotation of its quality. The genome annotation can happen afterwards.
2. right now, we are transitioning from the old way of having a special control probe field to treating control probes the same way as other regular probes. These probes are distinguished by simply the probe name. If you see "ctl-" instead of "cg" or "rs" then that's a control probe. Yes, your understanding is correct, the controls field will not be used for future arrays. It's just a legacy from the older HM450 and EPIC array where the control probes are released separately.

Best,

Wanding

**Comment by Pancreas-Pratik** [comment]
Thank you Professor Zhou,

I have one last question just to give me some peace of mind before continuing my analysis with the fresh mm39 ; )

For question 2, When you say transitioning from the old way, will the transition have any implications on doing the analysis through mm39 now? What I mean to say is that, will there be changes in the future that will impact the integrity of the analysis done now? Especially if I am doing mm39, something that hasn't been done before? Theoretically speaking, at least?

I genuinely appreciate your help again, sir.

Sincerely,
Pratik

**Comment by zwdzwd** [comment]
for question 2, I was referring to the way control probes are used. I actually don't recall they would be used in a genome-assembly-dependent setting. Currently in sesame, you can see that we have function to handle control probes in both the legacy way and as regular probes
https://github.com/zwdzwd/sesame/blob/a747c6ea64f4f2d429b8e190200dc5ad76c11cd6/R/sesame.R#L321
and its use
https://github.com/zwdzwd/sesame/blob/ad9abb915d19ef07f259ada8500e4eb79a960186/R/detection.R#L11
in short, no, I don't think the control probe format will affect mm39 analysis.

Wanding
